ambry-replication/src/main/java/com.github.ambry.replication/ReplicationEngine.java

```diff
@@ -0,0 +1,708 @@
+
+      int numberOfNodesPerThread = dataNodesCount / replicaThreadCount;
+      int remainingNodes = dataNodesCount % replicaThreadCount;
+
+      Iterator<DataNodeId> dataNodeIdIterator = dataNodesToReplicate.iterator();
+
+      for (int i = 0; i < replicaThreadCount; i++) {
+        // create the list of nodes for the replica thread
+        Map<DataNodeId, List<RemoteReplicaInfo>> replicasForThread = new HashMap<DataNodeId, List<RemoteReplicaInfo>>();
+        int nodesAssignedToThread = 0;
+        while (nodesAssignedToThread < numberOfNodesPerThread) {
+          DataNodeId dataNodeToReplicate = dataNodeIdIterator.next();
+          replicasForThread.put(dataNodeToReplicate,
+              dataNodeRemoteReplicaInfos.getRemoteReplicaListForDataNode(dataNodeToReplicate));
+          dataNodeIdIterator.remove();
+          nodesAssignedToThread++;
+        }
+        if (remainingNodes > 0) {
+          DataNodeId dataNodeToReplicate = dataNodeIdIterator.next();
+          replicasForThread.put(dataNodeToReplicate,
+              dataNodeRemoteReplicaInfos.getRemoteReplicaListForDataNode(dataNodeToReplicate));
+          dataNodeIdIterator.remove();
+          remainingNodes--;
+        }
+        boolean replicatingOverSsl = sslEnabledDatacenters.contains(datacenter);
+        String threadIdentity =
+            "Replica Thread-" + (dataNodeId.getDatacenterName().equals(datacenter) ? "Intra-" : "Inter") + i
+                + datacenter;
+        try {
+          StoreKeyConverter threadSpecificKeyConverter = storeKeyConverterFactory.getStoreKeyConverter();
+          Transformer threadSpecificTransformer =
+              Utils.getObj(transformerClassName, storeKeyFactory, threadSpecificKeyConverter);
+          ReplicaThread replicaThread =
+              new ReplicaThread(threadIdentity, replicasForThread, factory, clusterMap, correlationIdGenerator,
+                  dataNodeId, connectionPool, replicationConfig, replicationMetrics, notification,
+                  threadSpecificKeyConverter, threadSpecificTransformer, metricRegistry, replicatingOverSsl, datacenter,
+                  responseHandler, SystemTime.getInstance());
+          if (replicaThreadPools.containsKey(datacenter)) {
+            replicaThreadPools.get(datacenter).add(replicaThread);
+          } else {
+            replicaThreadPools.put(datacenter, new ArrayList<>(Arrays.asList(replicaThread)));
+          }
+        } catch (Exception e) {
+          throw new IOException("Encountered exception instantiating ReplicaThread", e);
+        }
+      }
+    }
+  }
+
+  /**
+   * Reads the replica tokens from the file and populates the Remote replica info
+   * and persists the token file if necessary.
+   * @param mountPath The mount path where the replica tokens are stored
+   * @throws ReplicationException
+   * @throws IOException
+   */
+  private void readFromFileAndPersistIfNecessary(String mountPath) throws ReplicationException, IOException {
```

Comment:
The tokens are combined  because their Partitions are under same mount path. For VCR case, we can set cloudPartition mountPath differently. 

If we really need to split, `partitionGroupedByMountPath` initialization will be changed too. We can do this in future PR.